addons/link_tracker/models/mail_render_mixin.py

```diff
@@ -34,29 +37,30 @@ def _shorten_links(self, html, link_tracker_vals, blacklist=None, base_url=None)
 
         :return: updated html
         """
+        if not html or is_html_empty(html):
+            return html
         base_url = base_url or self.env['ir.config_parameter'].sudo().get_param('web.base.url')
         short_schema = base_url + '/r/'
-        for match in set(re.findall(tools.HTML_TAG_URL_REGEX, html)):
-            long_url = match[1]
-            # Make relative links absolute
-            if long_url.startswith(('/', '?', '#')):
-                long_url = base_url + long_url
-            # Don't shorten already-shortened links
-            if long_url.startswith(short_schema):
-                continue
-            # Don't shorten urls present in blacklist (aka to skip list)
-            if blacklist and any(s in long_url for s in blacklist):
-                continue
-            label = (match[3] or '').strip()
 
-            create_vals = dict(link_tracker_vals, url=unescape(long_url), label=unescape(label))
-            link = self.env['link.tracker'].search_or_create(create_vals)
-            if link.short_url:
-                # `str` manipulation required to support replacing "&" characters, common in urls
-                new_href = match[0].replace(match[1], link.short_url)
-                html = html.replace(markupsafe.Markup(match[0]), markupsafe.Markup(new_href))
+        root_node = lxml.html.fromstring(html)
+        link_nodes, urls_and_labels = find_links_with_urls_and_labels(
+            root_node, base_url, skip_regex=rf'^{URL_SKIP_PROTOCOL_REGEX}', skip_prefix=short_schema,
+            skip_list=blacklist)
+        if not link_nodes:
+            return html
+
+        links_trackers = self.env['link.tracker'].search_or_create([
+            dict(link_tracker_vals, url=unescape(url), label=unescape(label))
```

Comment:
Shoudln't we unescape before calling "search_or_create" ? Unsure it is really the role of this "crud-like" method to do it